colossalai/nn/optimizer/distributed_lamb.py

```diff
@@ -0,0 +1,167 @@
+                    state["step"] = 0
+                    # Exponential moving average of gradient values
+                    state["exp_avg"] = torch.zeros_like(p.data)
+                    # Exponential moving average of squared gradient values
+                    state["exp_avg_sq"] = torch.zeros_like(p.data)
+
+                exp_avg, exp_avg_sq = state["exp_avg"], state["exp_avg_sq"]
+                beta1, beta2 = group["betas"]
+
+                state["step"] += 1
+
+                # Decay the first and second moment running average coefficient
+                # m_t
+                exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
+                # v_t
+                exp_avg_sq.mul_(beta2).addcmul_(grad, grad, value=1 - beta2)
+
+                scaled_lr = group["lr"]
+                if group["bias_correction"]:
+                    bias_correction1 = 1 - beta1 ** state["step"]
+                    bias_correction2 = 1 - beta2 ** state["step"]
+                    # Apply debiasing to lr to avoid broadcast
+                    scaled_lr *= (bias_correction2**0.5) / bias_correction1
+                    # exp_avg.div_(bias_correction1)
+                    # exp_avg_sq.div_(bias_correction2)
+
+                update = exp_avg / exp_avg_sq.sqrt().add(group["eps"])
+                if group["weight_decay"] != 0:
+                    update.add_(p.data, alpha=group["weight_decay"])
+
+                # Compute global layer-wise trust ratio
+                is_dist = (
+                    is_distributed_tensor(p)
+                    if self.dp_size <= 1
+                    else is_distributed_tensor(self.shard_to_param.get(id(p), None))
+                )
+
+                if is_dist:
+                    p_local = p
+                    g_sum = (update**2).sum()
+                    if self.dp_size > 1:
+                        # ZeRO 2 doesn't shard param. Compute full param norm w/o communication.
+                        dist.all_reduce(g_sum, group=self.dp_group)
+                        p_local = self.shard_to_param[id(p)]
+                    w_sum = (p_local**2).sum()
+                    sums = torch.stack([w_sum, g_sum])
+
+                    # Get global l2 norms
+                    if self.tp_size > 1:
+                        dist.all_reduce(sums, group=self.tp_group)
+                    w_norm, g_norm = sums.sqrt().chunk(2)
+                else:
+                    # Fall back to vanilla Lamb
+                    w_norm = torch.norm(p)
+                    g_norm = torch.norm(update)
+
+                trust_ratio = torch.where(
+                    w_norm > 0 and g_norm > 0, w_norm / g_norm, torch.tensor(1.0, device=w_norm.device)
+                )
+
+                state["weight_norm"] = w_norm
+                state["adam_norm"] = g_norm
+                state["trust_ratio"] = trust_ratio
+
+                scaled_lr *= trust_ratio
+                p.data.add_(update, alpha=-scaled_lr.item())
```

Comment:
DON'T use `.item()`, which may trigger D2H data movement

Reply:
add_ requires alpha to be non-tensor